pkg/dyninst/exprlang/exprlang.go

```diff
@@ -0,0 +1,158 @@
+// in the expression language of the dynamic instrumentation and live debugger products.
+package exprlang
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/go-json-experiment/json/jsontext"
+)
+
+// Expr represents an expression in the DSL.
+type Expr interface {
+	expr() // marker method
+}
+
+// RefExpr represents a reference to a variable.
+type RefExpr struct {
+	Ref string
+}
+
+func (re *RefExpr) expr() {}
+
+// UnsupportedExpr represents an expression type that is not yet supported.
+type UnsupportedExpr struct {
+	operation string
+	argument  jsontext.Value
+}
+
+func (ue *UnsupportedExpr) expr() {}
+
+var decoderPool = sync.Pool{
+	New: func() any {
+		r := &bytes.Reader{}
+		d := jsontext.NewDecoder(r)
+		return &pooledDecoder{decoder: d, reader: r}
+	},
+}
+
+type pooledDecoder struct {
+	decoder *jsontext.Decoder
+	reader  *bytes.Reader
+}
+
+func getPooledDecoder(data []byte) *pooledDecoder {
+	d := decoderPool.Get().(*pooledDecoder)
+	d.reader.Reset(data)
+	return d
+}
+
+func (d *pooledDecoder) put() {
+	d.reader.Reset(nil)
+	d.decoder.Reset(d.reader)
+	decoderPool.Put(d)
+}
+
+// Parse parses a DSL JSON expression into a strongly-typed AST node.
+func Parse(dslJSON []byte) (Expr, error) {
+	if len(dslJSON) == 0 {
+		return nil, fmt.Errorf("parse error: empty DSL expression")
+	}
+	pooled := getPooledDecoder(dslJSON)
+	defer pooled.put()
+	dec := pooled.decoder
+
+	// Ensure we have a JSON object
+	objStart, err := dec.ReadToken()
+	if err != nil {
+		return nil, fmt.Errorf("parse error: failed to read token: %w", err)
+	}
+	if kind := objStart.Kind(); kind != '{' {
+		return nil, fmt.Errorf("parse error: malformed DSL: got token %v (%v), expected {", objStart, kind)
+	}
+
+	// Read the operation key
+	key, err := dec.ReadToken()
+	if err != nil {
+		return nil, fmt.Errorf("parse error: failed to read operation key: %w", err)
+	}
+	if kind := key.Kind(); kind != '"' {
+		return nil, fmt.Errorf("parse error: malformed DSL: got key token %v (%v), expected string", key, kind)
+	}
+
+	operation := key.String()
+	switch operation {
+	case "ref":
+		// Read the ref value
+		val, err := dec.ReadToken()
+		if err != nil {
+			return nil, fmt.Errorf("parse error: failed to read ref value: %w", err)
+		}
+		if kind := val.Kind(); kind != '"' {
+			return nil, fmt.Errorf("parse error: malformed ref: got value token %v (%v), expected string", val, kind)
+		}
+
+		refValue := val.String()
+		if refValue == "" {
+			return nil, fmt.Errorf("parse error: ref value cannot be empty")
+		}
+
+		// Read the closing brace
+		endObj, err := dec.ReadToken()
+		if err != nil {
+			return nil, fmt.Errorf("parse error: failed to read closing brace: %w", err)
+		}
+		if kind := endObj.Kind(); kind != '}' {
+			return nil, fmt.Errorf("parse error: malformed DSL: got token %v (%v), expected }", endObj, kind)
+		}
+
+		return &RefExpr{Ref: refValue}, nil
+	default:
+		// Read the argument for unsupported operations
+		arg, err := dec.ReadToken()
```

Comment:
this is not correct, you don't want to read a token, you want to read a value: https://pkg.go.dev/github.com/go-json-experiment/json/jsontext#Decoder.ReadValue.

This thing where you call a `String()` on token and then cast it to jsontext.Value is a pattern that is roughly never going to be correct (note that for string values this gives you the unescaped string).

The more complex example tests I asked for would have clarified this.

Reply:
Another thing that I think will be cool is that we don't actually need to do a slices.Clone with the returned value because one thing we can do is use `InputOffset` before and after `ReadValue` and then subslice the input from there to avoid copying out any data.

Reply:
sorry, I got ahead of myself. If you just use `ReadValue` it's only valid until the next call on the decoder, so to store it you'd need to copy out the bytes.